From 23ced37b06c9c4af1dd8e71ce25d3c2c8cc40a8b Mon Sep 17 00:00:00 2001
From: Peter Swain <pswain@cavium.com>
Date: Tue, 3 Jul 2018 15:28:25 -0700
Subject: [PATCH 0017/1051] mtd: cavium_nand: power management

Added power management code, untested.
Not so much for immediate use of CONFIG_PM, but so the necessary
maintenance of iface_mode/iface_set, distinguishing NAND reset
events from NDF-controller PM events, is coded.

Signed-off-by: Peter Swain <pswain@cavium.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/mtd/nand/raw/cavium_nand.c | 44 ++++++++++++++++++++++++++++--
 1 file changed, 42 insertions(+), 2 deletions(-)

diff --git a/drivers/mtd/nand/raw/cavium_nand.c b/drivers/mtd/nand/raw/cavium_nand.c
index 75d490a981a5..689cce4a508c 100644
--- a/drivers/mtd/nand/raw/cavium_nand.c
+++ b/drivers/mtd/nand/raw/cavium_nand.c
@@ -1348,8 +1348,9 @@ static int cvm_nfc_chip_init(struct cvm_nfc *tn, struct device *dev,
 	return 0;
 }
 
-static int cvm_nfc_chips_init(struct cvm_nfc *tn, struct device *dev)
+static int cvm_nfc_chips_init(struct cvm_nfc *tn)
 {
+	struct device *dev = tn->dev;
 	struct device_node *np = dev->of_node;
 	struct device_node *nand_np;
 	int nr_chips = of_get_child_count(np);
@@ -1473,7 +1474,7 @@ static int cvm_nfc_probe(struct pci_dev *pdev,
 	}
 
 	cvm_nfc_init(tn);
-	ret = cvm_nfc_chips_init(tn, dev);
+	ret = cvm_nfc_chips_init(tn);
 	if (ret) {
 		dev_err(dev, "failed to init nand chips\n");
 		goto error;
@@ -1500,6 +1501,41 @@ static void cvm_nfc_remove(struct pci_dev *pdev)
 	clk_disable_unprepare(tn->clk);
 }
 
+#ifdef CONFIG_PM_SLEEP
+static int cvm_nfc_suspend(struct pci_dev *pdev, pm_message_t unused)
+{
+	struct cvm_nfc *tn = pci_get_drvdata(pdev);
+	struct cvm_nand_chip *chip;
+
+	list_for_each_entry(chip, &tn->chips, node)
+		chip->iface_set = false;
+	clk_disable_unprepare(tn->clk);
+
+	return 0;
+}
+
+static int cvm_nfc_resume(struct pci_dev *pdev)
+{
+	struct cvm_nfc *tn = pci_get_drvdata(pdev);
+	int ret = clk_prepare_enable(tn->clk);
+
+	if (ret) {
+		dev_err(tn->dev, "failed to enable clk\n");
+		return ret;
+	}
+
+	/* can some of this be skipped, or refactored... */
+	cvm_nfc_init(tn);
+	ret = cvm_nfc_chips_init(tn);
+	if (ret) {
+		dev_err(tn->dev, "failed to resume nand chips\n");
+		return ret;
+	}
+
+	return 0;
+}
+#endif
+
 static const struct pci_device_id cvm_nfc_pci_id_table[] = {
 	{ PCI_DEVICE(PCI_VENDOR_ID_CAVIUM, 0xa04f) },
 	{ 0, }
@@ -1512,6 +1548,10 @@ static struct pci_driver cvm_nfc_pci_driver = {
 	.id_table	= cvm_nfc_pci_id_table,
 	.probe		= cvm_nfc_probe,
 	.remove		= cvm_nfc_remove,
+#ifdef CONFIG_PM_SLEEP
+	.suspend	= cvm_nfc_suspend,
+	.resume		= cvm_nfc_resume,
+#endif
 };
 
 module_pci_driver(cvm_nfc_pci_driver);
-- 
2.17.1

